src/controllers/auth/protect.ts

```diff
@@ -0,0 +1,103 @@
+
+export const protect = catchAsync(async (req: Request, res: Response, next: NextFunction) => {
+	// get the cookies from the request headers
+	const { abegAccessToken, abegRefreshToken } = req.cookies as {
+		abegAccessToken: string;
+		abegRefreshToken: string;
+	};
+
+	if (!abegAccessToken || !abegRefreshToken) {
+		throw new AppError('Auth tokens not provided', 401);
+	}
+
+	// verify user access
+
+	const handleUserVerification = async (decoded) => {
+		// fetch user from redis cache or db
+		const cachedUser = (await getFromCache(decoded.id)) as IUser;
+		const user = cachedUser
+			? cachedUser
+			: ((await User.findOne({ _id: decoded.id })
+					.select('refreshToken loginRetries isSuspended isEmailVerified lastLogin')
+					.lean()) as IUser);
+
+		if (!cachedUser && user) {
+			await setCache(decoded.id, user);
+		}
+
+		// check if refresh token matches the stored refresh token in db
+		// in case the user has logged out and the token is still valid
+		// or the user has re authenticated and the token is still valid etc
+
+		if (user.refreshToken !== abegRefreshToken) {
+			throw new AppError('Invalid token. Please log in again!', 401);
+		}
+
+		if (user.isSuspended) {
+			throw new AppError('Your account is currently suspended', 401);
+		}
+
+		if (!user.isEmailVerified) {
+			throw new AppError('Your email is yet to be verified', 401);
+		}
+
+		// check if user has changed password after the token was issued
+		// if so, invalidate the token
+		if (
+			DateTime.fromISO(user.passwordChangedAt.toISOString()).toMillis() > DateTime.fromMillis(decoded.iat).toMillis()
+		) {
+			throw new AppError('Password changed since last login. Please log in again!', 401);
+		}
+
+		// csrf protection
+		// browser client fingerprinting
+
+		return user;
+	};
+
+	try {
+		const verifyAsync: (arg1: string, arg2: string) => jwt.JwtPayload = promisify(jwt.verify);
+		const decodeAccessToken = verifyAsync(abegAccessToken, ENVIRONMENT.JWT.ACCESS_KEY!);
+		const currentUser: IUser = await handleUserVerification(decodeAccessToken);
+
+		// attach the user to the request object
+		(req as CustomRequest).user = currentUser;
+	} catch (error) {
+		if (error instanceof jwt.JsonWebTokenError || error instanceof jwt.TokenExpiredError) {
+			// verify the refresh token and generate a new access token
+			try {
+				const verifyAsync = promisify<string, string>(jwt.verify);
+				const decodeRefreshToken = await verifyAsync(abegRefreshToken, ENVIRONMENT.JWT.REFRESH_KEY!);
+
+				const currentUser: IUser = await handleUserVerification(decodeRefreshToken);
+
+				const accessToken = jwt.sign({ id: currentUser._id }, ENVIRONMENT.JWT.ACCESS_KEY, {
+					expiresIn: JWTExpiresIn.Access,
+				});
+
+				setCookie(res, 'abegAccessToken', accessToken, {
+					maxAge: JWTExpiresIn.Access / 1000,
+				});
+
+				(req as CustomRequest).user = currentUser;
+			} catch (error) {
+				next(new AppError('Invalid token, please log in again', 500));
```

Comment:
500 status not supposed to be returned on purpose as it's only an indication of error | unhandled error in server

Reply:
Good catch, copilot at it again

Reply:
i have changed this in [this commit](https://github.com/abeg-help/backend/pull/89/commits/1a29bc51d428b489017e3d55ecc86316a3bc7be2`)